Support for Kafka 4.0 (#124)
strimzi/proposals: 076-support-for-kafka-4.0.md

   This would include things such as:
   * Deprecating the ZooKeeper related parts of the CRDs
   * Removing the ZooKeeper reconciler and all other parts of the code relevant to ZooKeeper
   * Remove the migration code
   * Remove ZooKeeper related system tests

   As upgrade to Strimzi 0.4z will not be possible without migration to KRaft, we will know for sure that the clusters run by Strimzi 0.4z had to be previously run by recent Strimzi versions.
   That will also allow us to remove other legacy code.
   For example the migration of Kafka clusters from StatefulSets to StrimziPodSets would not be needed anymore and would be removed.

   Once Apache Kafka 4.0 is released, we will adopt it and do a Strimzi release 0.4z.
   0.4z and all later Strimzi versions will support only KRaft based clusters.
   Strimzi 0.4z will support the last two Kafka minor versions as usually - Kafka 3.9 and 4.0 - and will allow users to gradually upgrade as always.
   But it will support only KRaft and users with ZooKeeper-based clusters will need to migrate to KRaft first within the Strimzi 0.4x and 0.4y versions and Kafka 3.9.

> 4) The Strimzi releases after 0.4z will continue as usually with the 1-2 months release cycles and will continue to adopt the Kafka patch releases.

Reviewer: ```suggestion
4) The Strimzi releases after 0.4z will continue as usually with the roughly 2 months release cycles, and will continue to adopt the Kafka patch releases.
```

Author: Our goal always was to release every 1-2 months.

In Strimzi 0.4z and newer versions we will check the Kafka clusters for the metadata type they use.
For any cluster not using KRaft it will immediately raise an error instead of trying to reconcile the cluster.
The check will be done based on the `Kafka` CR `.status` section.
Users will have to downgrade the operator, migrate the cluster and upgrade the operator again.

### Extended support for Kafka 3.9

Based on the feedback from the community, there seem to be some users who are very eager to move to KRaft.
But also many users who prefer to stick with ZooKeeper for longer.
We should try to accommodate these users by providing extended support for the last Strimzi release with support for Apache Kafka 3.9 and ZooKeeper-based clusters.
This extended support should include:
* Trying to fix critical CVEs
* Add support for new Apache Kafka 3.9 patch releases 
* Fixing critical bugs related to ZooKeeper, KRaft and migration from ZooKeeper to Kraft

The extended support will not include back-porting any new features.
It is also not expected to include any changes for compatibility with new Kubernetes versions released only after the Strimzi release.

> Releases of new patch versions will be not be done with a frequency higher than 2 months and should be driven based on user demand.

Author: It says that we will not do more frequent releases than every 2 months and we the main trigger for the releases will be user demand.

Unlike previous major releases that Strimzi handled before (2.0 and 3.0), this one is "special" because it drops the support for ZooKeeper-based clusters.
This proposal discusses how Strimzi will handle the release of Apache Kafka 4.0 and transition from Apache Kafka 3.x to 4.x.

## Current situation

### Strimzi

Strimzi currently does a new minor release every roughly every 2 months.
Each release supports at least two last Kafka minor versions.
For example 3.6.x and 3.7.x.
For each minor version, multiple patch releases might be supported, depending on their availability in Apache Kafka.
When new minor Apache Kafka version is released, Strimzi will typically adopt the new minor version and drop support for the oldest support Kafka minor version.
Typically, this happens every other Strimzi release as Apache Kafka has slower release cycle compared to Strimzi.
But in some cases, there might be more Strimzi releases in between (for example, Strimzi 0.40, 0.41 and 0.42 are all based on Kafka 3.6.x and 3.7.x).

Once a new minor Strimzi version is released, we will stop supporting the previous Strimzi release.
There are some exception, especially in case of:
* Serious bug or CVE

> * The latest minor Strimzi release is new and not anyone might have yet change to upgrade to is

Author: Yeah, that is pretty much what I meant by this. Fixed, thanks.

It should also help our users to properly plan when to migrate from ZooKeeper to KRaft, as this is a major change that might require planning several months ahead of time.

## Proposal

Strimzi should approach adoption of Kafka 4.0 in the following way:

1) When Apache Kafka 3.9.0 is released, we will adopt it in Strimzi as any other Apache Kafka release.
   We will release it as part of a Strimzi release 0.4x which will support both Kafka 3.8.x and 3.9.x.
   This version will support both ZooKeeper- and KRaft-based clusters.
   This version will also add warnings to all users using ZooKeeper that it will be removed in the next Kafka release and that they should do the migration.

2) Given the expected timeline of 3-4 months between Kafka 3.9 and 4.0, we will do another minor Strimzi release 0.4y.
   This version might add support for some of the new features added to KRaft in Kafka 3.9.0.
   0.4y will keep the support for 3.8 and 3.9 as well as support for both KRaft and ZooKeeper.
   This version will be subject to the extended support described below in a dedicated section.

>    It will also keep the warnings introduced in previous Strimzi version.

Reviewer 2: ```suggestion
   It will also keep the warnings introduced in Strimzi 0.4x.
```

3) Once Strimzi 0.4y is released, we will proceed and remove ZooKeeper support from Strimzi.
   This would include things such as:
   * Deprecating the ZooKeeper related parts of the CRDs
   * Removing the ZooKeeper reconciler and all other parts of the code relevant to ZooKeeper
   * Remove the migration code

>    * Remove ZooKeeper related system tests

Author: Correct. I added it to the list as well.

   As upgrade to Strimzi 0.4z will not be possible without migration to KRaft, we will know for sure that the clusters run by Strimzi 0.4z had to be previously run by recent Strimzi versions.
   That will also allow us to remove other legacy code.
   For example the migration of Kafka clusters from StatefulSets to StrimziPodSets would not be needed anymore and would be removed.

   Once Apache Kafka 4.0 is released, we will adopt it and do a Strimzi release 0.4z.
   0.4z and all later Strimzi versions will support only KRaft based clusters.
   Strimzi 0.4z will support the last two Kafka minor versions as usually - Kafka 3.9 and 4.0 - and will allow users to gradually upgrade as always.
   But it will support only KRaft and users with ZooKeeper-based clusters will need to migrate to KRaft first within the Strimzi 0.4x and 0.4y versions and Kafka 3.9.

> 4) The Strimzi releases after 0.4z will continue as usually with the 1-2 months release cycles and will continue to adopt the Kafka patch releases.

Reviewer 2: ```suggestion
4) The Strimzi releases after 0.4z will continue with the 1-2 months release cycles, as usual, and will continue to adopt the Kafka patch releases.
```

We will also remove the `kraft-regression` CI pipeline from the `main` branch and use the `regression` branch for running Kraft-based system tests.
The system test pipelines will remain unchanged in the 0.4y release branch for the extended support.
Given the extended support for the Strimzi 0.4y version, we should consider adding a dedicated upgrade test from a KRaft-based Strimzi 0.4y cluster to the latest Strimzi version as some users are expected to run Strimzi 0.4y for extended time and later migrate directly to a newer Strimzi version.

### Existing KRaft limitations

The existing KRaft limitations in Strimzi are caused by limitations in Apache Kafka itself.
While some of them might be addressed in Apache Kafka 3.9.0, additional time will be needed to adopt them in Strimzi and they might not be all addressed in the last Strimzi 0.4y release with ZooKeeper support.
They might be supported only in Strimzi 0.4z release based on Kafka 3.9 and 4.0 with KRaft-only support.
Support for these features will not be back-ported to the older Strimzi version with extended support.

### Risks

> Leaving aside the questions about the quality of the KRaft implementation and the missing features, the biggest risk are last-minute changes to the Apache Kafka release timelines and general plans.

Reviewer 2: ```suggestion
Leaving aside the questions about the quality of the KRaft implementation and the missing features, the biggest risks are last-minute changes to the Apache Kafka release timelines and general plans.
```

## Affected projects

This proposal affects the Strimzi operator.
Mainly the Cluster Operator which deals with KRaft and ZooKeeper.
But it to some extend affects also other parts of the project (such as the API, container images, examples, system tests etc.)
The Topic and User operators as well as the Bridge, Drain Cleaner or configuration providers should not be affected by this proposal as they do not have any direct dependency on ZooKeeper.

## Backwards compatibility

While this proposal maintains the backwards compatibility for existing KRaft clusters, it requires all users to migrate the ZooKeeper-based Kafka clusters to KRaft first.
This is not a Strimzi's choice, but a result of the development in the Apache Kafka project.

## Rejected alternatives

### Skip the Strimzi 0.4y release

We could consider skipping the Strimzi 0.4y release and instead remove the support for ZooKeeper-based clusters right after releasing Strimzi 0.4x.
In that case, the Strimzi 0.4x will receive the extended support.

> But assuming the interval between Kafka 3.9 and 4.0 is 3-4 months, it seems that keeping the 0.4y release can help us implemented some more features and possibly fix some more bug fixes before we drop ZooKeeper support.

Reviewer 2: ```suggestion
But assuming the interval between Kafka 3.9 and 4.0 is 3-4 months, it seems that keeping the 0.4y release can help us implement some more features and possibly fix some more bug fixes before we drop ZooKeeper support.
```

So it would leave us possibly for a very long time without any release and does not seem to bring any real advantages over the approach suggested by this proposal.

### Continue ZooKeeper support until Kafka 4.1 is released

Another alternative option would be to continue releasing Strimzi versions based on Apache Kafka 3.9 and 4.0 until Kafka 4.1 is released while supporting ZooKeeper with Kafka 3.9.
Once Kafka 4.1 would be released, we would drop support for Kafka 3.9 and for ZooKeeper and continue with Strimzi versions supporting Kafka 4.0 and 4.1 with KRaft only.

The disadvantage of this approach is that it would require additional effort to validate the used versions and make sure that users are not trying to deploy the ZooKeeper based Kafka cluster with Kafka 4.0.
It might also lead to confusing user experience.
For example, a Kafka CR without explicitly specified Kafka version should use the latest supported Kafka version.
Would that mean that ZooKeeper based clusters without specified `.spec.kafka.version` would automatically end up in error state because they would use Kafka 4.0?

> Or will we add some special handling that in such cases the version would default to Kafka 3.9?

Reviewer 3: I think this should be easily handled by some boolean or? If ZK is used then use only Kafka 3.9, and if someone wants to deploy with a different version i.e., 4.0+ then throw an error.

Author: I do not think it would be a simple boolean.

# Support for Apache Kafka 4.0

Apache Kafka is planning a new major release - Apache Kafka 4.0.

> Unlike previous major releases that Strimzi handled before (2.0 and 3.0), this one is "special" because it drops the support for ZooKeeper-based clusters.

Reviewer 4: ```suggestion
Unlike previous major releases that Strimzi handled before (2.x and 3.x), this one is "special" because it drops the support for ZooKeeper-based clusters.
```

This proposal discusses how Strimzi will adopt the release of Apache Kafka 4.0 and transition from Apache Kafka 3.x to 4.x.

## Current situation

### Strimzi

> Strimzi currently does a new minor release every roughly every 2 months.

Reviewer 4: ```suggestion
Strimzi currently does a new minor release roughly every 2 months.
```

When a new minor Apache Kafka version is released, Strimzi will typically adopt the new minor version and drop support for the oldest supported minor Kafka version.
Typically, this happens every other Strimzi release as Apache Kafka has slower release cycles compared to Strimzi.
But in some cases, there might be more Strimzi releases in between (for example, Strimzi 0.40, 0.41 and 0.42 are all based on Kafka 3.6.x and 3.7.x).

Once a new minor Strimzi version is released, we will stop supporting the previous Strimzi release.
There are some exceptions, especially in the following cases:
* Serious bugs or CVEs
* The latest minor Strimzi release is new and some people might not have had chance to upgrade to it

#### KRaft support

Strimzi currently supports both ZooKeeper- and KRaft-based Apache Kafka clusters.
It also supports migration of ZooKeeper-based clusters to KRaft.
The current limitations to the KRaft support are based on limitations in Apache Kafka itself.
At the time of writing this proposal, limitations include:
* Scaling of KRaft controller nodes up or down is not supported

> * Unregistering Kafka nodes removed from the Kafka cluster

Reviewer 5: In case some users read this proposal but doesn't have context what we are talking about here, we could add KIP/JIRA number:

```
* Scaling of KRaft controller nodes up or down is not supported (KIP-853)
* Unregistering Kafka nodes removed from the Kafka cluster (KAFKA-17094)
```

* Apache Kafka 3.9.0 release with the remaining missing KRaft features (shorter release cycle - expected August or September?)
* Apache Kafka 4.0.0 release 3-4 months after the Kafka 3.9 release without ZooKeeper support (late 2024, early 2025?)

The release plan is of course subject to change.

Apache Kafka currently supports every minor release for 1 year.
During this time, each minor release would typically get one or two patch releases with bug or CVE fixes.
The 1 year period is currently expected to apply for the Apache Kafka 3.9 release that is expected to be the last with ZooKeeper support.

## Motivation

The move from Apache Kafka 3.x to 4.x and the related move from ZooKeeper-based Kafka clusters to KRaft is a major change for our users as well as for Strimzi itself.
Having a good plan for how we plan to deal with it is important as it will provide clarity to Strimzi users, developers, and create a plan of the things that need to be done.
It should also help our users to properly plan when to migrate from ZooKeeper to KRaft, as this is a major change that might require planning several months ahead of time.

> ## Proposal

Reviewer 4: While the paragraph is well written and totally understandable I was wondering if a table would be useful at the end as a kind of summary and a way to have the big picture in one shot.
You could have one line for each Strimzi release 0.4x, 0.4y and 0.4z.
Then columns specifying: Kafka versions supported, metadata supported (ZooKeeper and/or KRaft).

Author: Done

Reviewer 4: I love it! Thanks!

Strimzi should approach adoption of Kafka 4.0 in the following way:

1) When Apache Kafka 3.9.0 is released, we will adopt it in Strimzi as any other Apache Kafka release.
   We will release it as part of a Strimzi release 0.4x which will support both Kafka 3.8.x and 3.9.x.
   This version will support both ZooKeeper- and KRaft-based clusters.

>    This version will also add warnings to all users using ZooKeeper that it will be removed in the next Kafka release and that they should do the migration.

Reviewer 4: the "warnings" you are mentioning are just in the cluster operator log or on the Kafka CR? maybe you can clarify it?

Author: They will be in both log and status. Added to the proposal.

   Once Apache Kafka 4.1 is released, Strimzi will adopt Kafka 4.1 and continue its releases with support for Kafka 4.0 and 4.1.

### Enforcing the migration to KRaft

There is no way we can enforce the migration to KRaft or do it automatically.
Users have to create the controller nodes and initiate the migration.
This cannot be done automatically, as we would not know what the right configuration for the KRaft controller should be (storage, resources, affinity, tolerations, ...).

To try to get users to migrate to KRaft in time, we will produce warning log messages and warning conditions in the `.status` section in the last Strimzi version supporting ZooKeeper (0.4x and 0.4y).

In Strimzi 0.4z and newer versions we will check the Kafka clusters for the metadata type they use.
For any cluster not using KRaft it will immediately raise an error instead of trying to reconcile the cluster.
The check will be done based on the `Kafka` CR `.status` section.
Users will have to downgrade the operator, migrate the cluster and upgrade the operator again.

Newly deployed Kafka clusters will be expected to be KRaft only.

> ### Extended support for Kafka 3.9

Reviewer 4: can you add in this paragraph that the Strimzi version with this support will be the 0.4y release that you mentioned before? Just to make it clearer.

Based on the feedback from the community, there seem to be some users who are very eager to move to KRaft.
But also many users who prefer to stick with ZooKeeper for longer.
We should try to accommodate these users by providing extended support for the last Strimzi release with support for Apache Kafka 3.9 and ZooKeeper-based clusters.
This extended support should include:
* Trying to fix critical CVEs
* Add support for new Apache Kafka 3.9 patch releases 
* Fixing critical bugs related to ZooKeeper, KRaft and migration from ZooKeeper to Kraft

The extended support will not include back-porting any new features.
It is also not expected to include any changes for compatibility with new Kubernetes versions released only after the Strimzi release.

> Releases of new patch versions will be not be done with a frequency higher than 2 months and should be driven based on user demand.

Reviewer 6: ```suggestion
Releases of new patch versions will not be done with a frequency higher than 2 months and should be driven based on user demand.
```

The extended support will be done at most for as long as the Kafka 3.9 version is supported by the Apache Kafka project (i.e. 1 year after the Kafka 3.9 release).
But it might be terminated sooner in case we run into major issues with fixing any bugs or CVEs that would require unreasonably high effort.
(While it is rare, addressing certain CVEs might require significant dependency updates that require major changes to the codebase.)

### System tests

With the removal of the ZooKeeper support from the production code, we should also proceed and remove it from the system tests.

> We will also remove the `kraft-regression` CI pipeline from the `main` branch and use the `regression` branch for running Kraft-based system tests.

Reviewer 4: should we also remove the `migration` pipeline?
